posts/2020-11-27-typing-your-actors.adoc

```diff
@@ -0,0 +1,389 @@
+      val dataProbe           = testKit.createTestProbe[Room.Response]()
+      val actingUserId        = UUID.randomUUID()
+      val (roomId, roomRef) = createRoom(
+        UUID.randomUUID(),
+        RoomData.empty.copy(users = List(user, user2))
+      )
+
+      val expectedMessage = WSMessage(MessageType.EditIssue, roomId, actingUserId, issue)
+      val expectedData = Room.DataStatus(data =
+        RoomData(
+          users = List(user, user2),
+          currentIssue = issue,
+          issueLastEditBy = Option(actingUserId)
+        )
+      )
+
+      roomRef ! Room.EditIssue(actingUserId, issue)
+
+      roomRef ! Room.GetData(dataProbe.ref)
+
+      userProbe.expectMsg(expectedMessage)
+      user2Probe.expectMsg(expectedMessage)
+
+      dataProbe.expectMessage(expectedData)
+    }
+  }
+----
+
+Takeaways:
+
+* Since actors are `Behavior` functions overriding internal functions for testing is not suggested like in https://doc.akka.io/docs/akka/current/testing.html#overriding-behavior[classical].
+* `BehaviorTestKit` is better at dealing with an actor's children.
+* `ActorTestKit` provides a simple way to create and use probes.
+* There isn't a clear way to test communication between more than 2 actors (if they aren't related).
+
+Tips:
+
+* Normally I would only expose the initial behavior from my actor, but I changed the other behavior access to private on the package, so now I can set the behavior data as I need before the test.
+[source,scala]
+----
+private[actors] def receiveBehaviour(roomId: UUID, data: RoomData): Behavior[Command] = ???
+----
+* Include a message that allow to inspect internal state, again access is private on package.
+[source,scala]
+----
+  private[actors] final case class GetData(replyTo: ActorRef[Response]) extends Command
+
+  sealed trait Response
+  final case class DataStatus(data: RoomData) extends Response
+----
+
+== Closing thoughts
+
+The migration went smoother than I was expecting. Once I started it I couldn't run the application until it was complete, and after changing the whole engine of the app I only faced two compilation errors that were quickly solved and the application was running again. Those errors were actually already present at in the application before the migration, however they went unnoticed due to not having the checks that typed actors bring, so the migration already proved useful.
+
+Some of the smoothness while migrating might be caused by already following good practices when using actors, specifically having a DSL already defined. If you're having more troubles when migrating, I would suggest to take a step back and refine your actors' DSL and transitions (How they communicate with each and states that they pass on their lifecycle).
+
+If you want to see more what could be done with typed actors and scala 3.0, https://blog.lunatech.com/2020/02/using-dotty-union-types-with-akka-typed/[here] is another blog it might interest you.
```

Comment:
Due to changes in the blog website, the link was broken. I've added a second article as the blog post was in two parts.
```suggestion
If you want to see more what could be done with typed actors and Scala 3.0, these two blog posts may be interesting to read: https://blog.lunatech.com/posts/2020-02-12-using-dotty-union-types-with-akka-typed[Using Scala 3 Union types with Akka Typed - part 1] and https://blog.lunatech.com/posts/2020-02-19-using-dotty-union-types-with-akka-typed-part-II[Using Scala 3 Union types with Akka Typed - part 2].
```